test/regress_listener.c

```diff
@@ -430,6 +445,94 @@ regress_listener_disable_in_thread_error(void *arg)
 }
 #endif
 
+#ifdef EVENT__HAVE_STRUCT_SOCKADDR_UN
+
+static void
+empty_listener_cb(struct evconnlistener *listener, evutil_socket_t fd,
+    struct sockaddr *sa, int socklen, void *user_data)
+{
+	printf("Empty listener, do nothing about it!\n");
+}
+
+static void
+regress_listener_reuseport_on_unix_socket(void *arg)
+{
+	struct basic_test_data *data = arg;
+	struct event_base *base = data->base;
+	struct evconnlistener *listener = NULL;
+	char *socket_path = NULL;
+	int truncated_temp_file = 0;
+	struct sockaddr_un addr;
+	int tempfd;
+
+	tempfd = regress_make_tmpfile("", 0, &socket_path);
+	/* On Windows, if TMP environment variable is corrupted, we may not be
+	 * able create temporary file, just skip it */
+	if (tempfd < 0)
+		tt_skip();
+	TT_BLATHER(("Temporary socket path: %s, fd: %i", socket_path, tempfd));
+
+#if defined(_WIN32) && defined(EVENT__HAVE_AFUNIX_H)
+	if (evutil_check_working_afunix_() == 0)
+		/* AF_UNIX is not available on the current Windows platform,
+		 * just skip this test. */
+		tt_skip();
+#endif
+
+	/* close fd to avoid leaks, since we only need a temporary file path
+	 * for bind() to create a socket file. */
+	close(tempfd);
+	/* For security reason, we must delete any existing sockets in the filesystem.
+	 * Besides, we will get EADDRINUSE error if the socket file already exists. */
+	unlink(socket_path);
+
+	memset(&addr, 0, sizeof(addr));
+	addr.sun_family = AF_UNIX;
+	if (strlcpy(addr.sun_path, socket_path, sizeof(addr.sun_path)) >= sizeof(addr.sun_path))
+		/* The temporary path exceeds the system limit, we may need to adjust some subsequent
+		 * assertions accordingly to ensure the success of this test. */
+		truncated_temp_file = 1;
+
+	listener = evconnlistener_new_bind(base, empty_listener_cb, (void *)base,
+	    LEV_OPT_REUSEABLE|LEV_OPT_CLOSE_ON_FREE, -1,
+			(struct sockaddr*)&addr, sizeof(addr));
+	tt_assert_msg(listener == NULL, "AF_UNIX listener shouldn't use SO_REUSEADDR!");
+
+	listener = evconnlistener_new_bind(base, empty_listener_cb, (void *)base,
+	    LEV_OPT_REUSEABLE_PORT|LEV_OPT_CLOSE_ON_FREE, -1,
+			(struct sockaddr*)&addr, sizeof(addr));
+	tt_assert_msg(listener == NULL, "AF_UNIX listener shouldn't use SO_REUSEPORT!");
+
+	/* Create a AF_UNIX listener without reusing address or port. */
+	listener = evconnlistener_new_bind(base, empty_listener_cb, (void *)base,
+	    LEV_OPT_CLOSE_ON_EXEC|LEV_OPT_CLOSE_ON_FREE, -1,
+			(struct sockaddr*)&addr, sizeof(addr));
+	if (truncated_temp_file) {
+		tt_assert_msg(listener == NULL, "Should not create a AF_UNIX listener with truncated socket path!");
+	} else {
+#ifdef _WIN32
+		if (listener == NULL && WSAENETDOWN == EVUTIL_SOCKET_ERROR())
```

Comment:
I hate how it looks, let's at least add a check that the temporary directory is unicode.